src/full/Agda/Termination/TypeBased/Checking.hs

```diff
@@ -0,0 +1,480 @@
+      ["Attempt to compare incomparable generic variables:"
+      , "t1: " <+> pretty t1
+      , "t2: " <+> pretty t2
+      ]
+    __IMPOSSIBLE__
+smallerOrEq pol (SizeArrow d1 c1) (SizeArrow d2 c2) = smallerOrEq (neg pol) d1 d2 >> smallerOrEq pol c1 c2 -- note the contravariance in domain
+smallerOrEq pol (SizeGeneric _ rest1) (SizeGeneric _ rest2) = smallerOrEq pol rest1 rest2
+smallerOrEq pol (SizeGeneric _ rest1) (SizeArrow UndefinedSizeType rest2) = smallerOrEq pol (sizeInstantiate UndefinedSizeType rest1) rest2
+smallerOrEq pol (SizeArrow UndefinedSizeType rest1) (SizeGeneric _ rest2) = smallerOrEq pol rest1 (sizeInstantiate UndefinedSizeType rest2)
+smallerOrEq pol (SizeTree _ _ _) (SizeArrow _ _) = pure () -- can occur, becase encoding of terms is intentionaly not complete
+smallerOrEq pol (SizeArrow _ _) (SizeTree _ _ _) = pure ()
+smallerOrEq pol (SizeArrow _ r) (SizeGenericVar args i) = smallerOrEq pol r (SizeGenericVar (args + 1) i) -- eta-conversion
+smallerOrEq pol (SizeGenericVar args i) (SizeArrow _ r) = smallerOrEq pol (SizeGenericVar (args + 1) i) r -- eta-conversion
+smallerOrEq pol t1 t2 = do
+  -- One example of a problem is an attempt to compare generic var and size variable.
+  -- This is an internal error, because it means that there is a forgotten instantiation somewhere.
+  reportSDoc "term.tbt" 20 $ vcat
+    ["Attempt to compare incomparable terms:"
+    , "t1: " <+> pretty t1
+    , "t2: " <+> pretty t2
+    , "polarity: " <+> pretty pol
+    ]
+  __IMPOSSIBLE__
+
+
+-- | Retrieves sized type for a constant
+-- May return Nothing for primitive definition
+resolveConstant :: QName -> TBTM ([SConstraint], SizeType)
+resolveConstant nm = do
+  reportSDoc "term.tbt" 20 $ "Resolving constant" <+> prettyTCM nm
+  currentName <- currentCheckedName
+  constInfo <- getConstInfo nm
+  sizedSig <- if currentName == nm
+    then do
+      let currentType = defType constInfo
+      liftTCM $ encodeFunctionType currentType
+    else pure $ defSizedType constInfo
+  let startingPolarity = case theDef constInfo of
+        Function {} -> Covariant
+        _ -> Invariant
+  freshenSignature startingPolarity sizedSig
+
+-- | Record information about a recursive call from current function to q2
+--   Only the calls withing the same mutual block matter.
+storeCall :: QName -> [Int] -> [Int] -> Elims -> TBTM ()
+storeCall q2 sizesq1 sizesq2 elims = do
+  names <- currentMutualNames
+  when (q2 `Set.member` names) do
+    currentName <- currentCheckedName
+    reportSDoc "term.tbt" 10 $ vcat
+      [ "Detected mutual-recursive call"
+      , "  From '" <> prettyTCM currentName <> "' with size variables: " <> text (show sizesq1)
+      , "  To   '" <> prettyTCM q2 <> "' with size variables: " <> text (show sizesq2)
+      ]
+    doc <- buildRecCallLocation q2 elims
+    reportCall q2 sizesq1 sizesq2 doc
+
+unwrapSizeTree :: SizeType -> [SizeType]
+unwrapSizeTree (SizeTree _ _ ts) = map snd ts
+unwrapSizeTree t = __IMPOSSIBLE__
```

Comment:
I have some code that explodes here. Unsure if the assumption is wrong or there's something else in the way. Will try to minimize it before friday.

Reply:
```
{-# OPTIONS --safe --cubical --type-based-termination --no-syntax-based-termination #-}
module TBTImpossible where

open import Agda.Primitive
open import Agda.Builtin.Sigma
open import Agda.Builtin.Cubical.Equiv

module _ {A B : Set} {f : A → B} where

  go : isEquiv f
  go .equiv-proof y .fst = {!!} -- C-c here
  go .equiv-proof y .snd = {!!}
```

Reply:
Sorry, I misguided you, there are two scenarios:
1. The culprit is `Patterns:238`, triggered by `C-c C-c` 
2. `C-c C-r` hits `Checking:464` later when trying to debug print something.

Seems like in the former case `ProjP` falls through a catch-all.

Reply:
Thanks for the reporting, these problems are fixed